dot/services_test.go

```diff
@@ -0,0 +1,656 @@
+			if tt.expectNil {
+				assert.Nil(t, got)
+			} else {
+				assert.NotNil(t, got)
+				assert.IsType(t, &system.Service{}, got)
+			}
+		})
+	}
+}
+
+func Test_newInMemoryDB(t *testing.T) {
+	tests := []struct {
+		name      string
+		path      string
+		expectNil bool
+		err       error
+	}{
+		{
+			name:      "working example",
+			path:      "test_data",
+			expectNil: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := newInMemoryDB(tt.path)
+			assert.ErrorIs(t, err, tt.err)
+
+			if tt.expectNil {
+				assert.Nil(t, got)
+			} else {
+				assert.NotNil(t, got)
+			}
+		})
+	}
+}
+
+//go:generate mockgen -destination=mock_babe_builder_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/lib/babe ServiceIFace,ServiceBuilder
+
+func Test_nodeBuilder_createBABEService(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockBabeBuilder := NewMockServiceBuilder(ctrl)
+	mockBabeIFace := NewMockServiceIFace(ctrl)
+	mockBabeBuilder.EXPECT().NewServiceIFace(gomock.AssignableToTypeOf(&babe.ServiceConfig{})).DoAndReturn(func(cfg *babe.
+		ServiceConfig) (babe.ServiceIFace, error) {
+		if reflect.ValueOf(cfg.BlockState).Kind() == reflect.Ptr && reflect.ValueOf(cfg.BlockState).IsNil() {
+			return nil, babe.ErrNilBlockState
+		}
+		return mockBabeIFace, nil
+	}).Times(2)
+	t.Parallel()
+	stateSrvc := newStateService(t)
+
+	cfg := NewTestConfig(t)
+
+	ks := keystore.NewGlobalKeystore()
+	ks2 := keystore.NewGlobalKeystore()
+	kr, err := keystore.NewSr25519Keyring()
+	require.NoError(t, err)
+	ks2.Babe.Insert(kr.Alice())
+
+	type args struct {
+		cfg             *Config
+		st              *state.Service
+		ks              keystore.Keystore
+		cs              *core.Service
+		telemetryMailer telemetry.Client
+	}
+	tests := []struct {
+		name     string
+		args     args
+		expected babe.ServiceIFace
+		err      error
+	}{
+		{
+			name: "invalid keystore",
+			args: args{
+				cfg: cfg,
+				st:  stateSrvc,
+				ks:  ks.Gran,
+			},
+			expected: nil,
+			err:      ErrInvalidKeystoreType,
+		},
+		{
+			name: "empty keystore",
+			args: args{
+				cfg: cfg,
+				st:  stateSrvc,
+				ks:  ks.Babe,
+			},
+			expected: nil,
+			err:      ErrNoKeysProvided,
+		},
+		{
+			name: "config error",
+			args: args{
+				cfg: cfg,
+				st:  &state.Service{},
+				ks:  ks2.Babe,
+			},
+			expected: nil,
+			err:      babe.ErrNilBlockState,
+		},
+		{
+			name: "base case",
+			args: args{
+				cfg: cfg,
+				st:  stateSrvc,
+				ks:  ks2.Babe,
+			},
+			expected: mockBabeIFace,
+			err:      nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			no := nodeBuilder{}
+			got, err := no.createBABEServiceBuilder(tt.args.cfg, tt.args.st, tt.args.ks, tt.args.cs,
+				tt.args.telemetryMailer, mockBabeBuilder)
+			assert.Equal(t, tt.expected, got)
+			assert.ErrorIs(t, err, tt.err)
+		})
+	}
+}
+
+func newStateService(t *testing.T) *state.Service {
```

Comment:
Check each usage of `newStateService` should be inside the subtest since it defines mocks.
You could also make it more explicit and take in a `ctrl *gomock.Controller` as second argument here, to also dedup having multiple gomock controllers.

Reply:
Updated to move newStateService into subtests and add gomock.Controller argument.